### PR TITLE
CUDA11 builders for Windows

### PR DESCRIPTION
**Pull request by zasdfgbnm**
cc: @peterjc123 https://github.com/pytorch/builder/pull/450 is deprecated and your comments in https://github.com/pytorch/builder/pull/450 will be resolved here. It is now WIP, not complete yet, and I don't have a chance to test it yet.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
@peterjc123 How do I debug these scripts? Do I just create azure pipelines on my fork with `windows/azure-pipelines.yml` and it will run everything?

**Comment by peterjc123** [comment]
We use CircleCI instead of Azure Pipelines now. I guess you'll need to modify https://github.com/pytorch/builder/blob/master/.circleci/regenerate.py. You can test that by removing the filter conditions in the builder jobs at PyTorch.

File: windows/internal/smoke_test.bat [hunk 1]
```diff
@@ -3,6 +3,7 @@ set SRC_DIR=%~dp0
 pushd %SRC_DIR%\..
 
 if "%CUDA_VERSION%" == "102" call internal\driver_update.bat
+if "%CUDA_VERSION%" == "110" call internal\driver_update.bat
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:6]: Have you checked that the driver is up to date?

File: windows/internal/cuda_install.bat [hunk 2]
```diff
@@ -87,6 +88,21 @@ if not exist "%SRC_DIR%\temp_build\cudnn-10.2-windows10-x64-v7.6.5.32.zip" (
...
+
+if not exist "%SRC_DIR%\temp_build\cuda_11.0.1_451.22_win10.exe" (
+    curl -k -L http://developer.download.nvidia.com/compute/cuda/11.0.1/local_installers/cuda_11.0.1_451.22_win10.exe --output "%SRC_DIR%\temp_build\cuda_11.0.1_451.22_win10.exe"
+    if errorlevel 1 exit /b 1
+    set "CUDA_SETUP_FILE=%SRC_DIR%\temp_build\cuda_11.0.1_451.22_win10.exe"
+    set "ARGS=nvcc_11.0 cuobjdump_11.0 nvprune_11.0 cupti_11.0 cublas_11.0 cublas_dev_11.0 cudart_11.0 cufft_11.0 cufft_dev_11.0 curand_11.0 curand_dev_11.0 cusolver_11.0 cusolver_dev_11.0 cusparse_11.0 cusparse_dev_11.0 nvgraph_11.0 nvgraph_dev_11.0 npp_11.0 npp_dev_11.0 nvrtc_11.0 nvrtc_dev_11.0 nvml_dev_11.0"
+)
+
+if not exist "%SRC_DIR%\temp_build\cudnn-11.0-windows-x64-v8.0.0.180.zip" (
+    curl -k -L https://developer.download.nvidia.com/compute/redist/cudnn/v8.0.0/cudnn-11.0-windows-x64-v8.0.0.180.zip --output "%SRC_DIR%\temp_build\cudnn-11.0-windows-x64-v8.0.0.180.zip"
```
Comment (peterjc123) [inline windows/internal/cuda_install.bat:101]: This link requires the credentials. So we have to upload it to S3.

File: windows/internal/smoke_test.bat (hunk 1 above)
Comment (zasdfgbnm) [reply windows/internal/smoke_test.bat:6]: Thanks for reminding me, and you are right that it is not up to date. It requires >=451.22. I will update this.

**Comment by robert-baldwin** [comment]
I'm just getting started with pytorch on Ubuntu on WSL2. Will this allow me to `conda install pytorch torchvision cudatoolkit=11.0 -c pytorch`?

**Comment by peterjc123** [comment]
> I'm just getting started with pytorch on Ubuntu on WSL2. Will this allow me to `conda install pytorch torchvision cudatoolkit=11.0 -c pytorch`?

Nope, I don't think there will be binaries for CUDA 11 recently because the conda package `cudatoolkit=11.0` is not released yet. But you may eventually be able do that no matter you are using WSL or not on Windows.

**Comment by robert-baldwin** [comment]
Understood, thank you. I have one follow-up question if you don't mind.

I tried cloning the pytorch repository and building from source from the `master` branch which successfully built and loaded in `python`, but `torch.cuda.is_available() #=> False` although prior to running `python setup.py install` I had `conda install -c nvidia cudatoolkit` which gave me version `11.0.171`. Would this branch potentially fix the local build issue?

**Comment by zasdfgbnm** [comment]
@robert-baldwin No, this fix is for the nightly builder, it is unrelated to what you are seeing. But what you are seeing is strange, there might be a problem with PyTorch somewhere.

**Comment by zasdfgbnm** [comment]
@robert-baldwin Are you building PyTorch inside WSL? I think GPU is not supported inside WSL in current Windows release. According to [Microsoft](https://devblogs.microsoft.com/commandline/gpu-compute-wsl-install-and-wsl-update-arrive-in-the-windows-insiders-fast-ring-for-the-windows-subsystem-for-linux/), it is only available on Windows Insider Preview Build 20150:
> There are three thrilling new updates for the Windows Subsystem for Linux (WSL) in the new Windows Insider Preview Build 20150. The first is GPU compute: a feature that allows your Linux binaries to leverage your GPU, which makes it possible to do more machine learning/AI development and data science workflows directly in WSL.

**Comment by robert-baldwin** [comment]
I appreciate the follow up @zasdfgbnm ! I am on the Windows Insider Preview Build 20152.1000 using Ubuntu 18.04 on WSL2 and have followed NVidia's CUDA Toolkit Documentation here: https://docs.nvidia.com/cuda/wsl-user-guide/index.html

I've verified that my GPU is reachable by watching Windows Task Manager while running scripts from the Cuda Samples in https://github.com/NVIDIA/cuda-samples

One thing I've been uncertain about is whether installing CUDA Toolkit via `apt` as outlined in NVidia's somehow interferes w/ the version installed via `conda` -- I suspect so, since I saw "cycles in the constraint graph" warnings when I had both installed -- and opted to run the pytorch source build w/ only `conda` installed version only.

At this point I'm not quite sure where to look to discover whether where my issue may be.

**Comment by robert-baldwin** [comment]
I'm also happy to open an issue or move this conversation to a more appropriate place to get support based on your guidance. I very much appreciate your time and am conscious about unintentionally monopolizing it 😄

**Comment by zasdfgbnm** [comment]
@robert-baldwin I actually don't know, because I haven't tried WSL yet. But if I were you, I would try
- nightly build, and
- compile from source with CUDA 10.2

**Comment by peterjc123** [comment]
> Understood, thank you. I have one follow-up question if you don't mind.
> 
> I tried cloning the pytorch repository and building from source from the `master` branch which successfully built and loaded in `python`, but `torch.cuda.is_available() #=> False` although prior to running `python setup.py install` I had `conda install -c nvidia cudatoolkit` which gave me version `11.0.171`. Would this branch potentially fix the local build issue?

No, it is only a runtime package, not a source package.

File: windows/internal/cuda_install.bat (hunk 2 above)
Comment (zasdfgbnm) [reply windows/internal/cuda_install.bat]: I updated this link to use the latest CUDA 11 and cuDNN release. These links should be working without login, but feel free to upload to S3 to keep consistent with other links.

**Comment by zasdfgbnm** [comment]
@peterjc123 This PR is merged now https://github.com/pytorch/builder/pull/461, could you please use this to build magma for me? Thanks :)

File: windows/internal/smoke_test.bat (hunk 1 above)
Comment (zasdfgbnm) [reply windows/internal/smoke_test.bat:6]: I believe I will need your help to download the latest driver and upload it to S3 so I can use the link :)

**Comment by zasdfgbnm** [comment]
This PR is tested by https://github.com/pytorch/pytorch/pull/42482

**Comment by peterjc123** [comment]
You could refer to the PR I mentioned below. Also magma is already built. See https://github.com/pytorch/pytorch/pull/42420

**Comment by zasdfgbnm** [comment]
@peterjc123 WOW, amazing! I was not aware of that PR. Thanks for working on it!

File: windows/internal/cuda_install.bat [hunk 3]
```diff
@@ -91,14 +91,14 @@ if not exist "%SRC_DIR%\temp_build\cudnn-10.2-windows10-x64-v7.6.5.32.zip" (
...
-    curl -k -L http://developer.download.nvidia.com/compute/cuda/11.0.2/local_installers/cuda_11.0.2_451.48_win10.exe --output "%SRC_DIR%\temp_build\cuda_11.0.2_451.48_win10.exe"
+    curl -k -L https://ossci-windows.s3.amazonaws.com/cuda_11.0.2_451.48_win10.exe --output "%SRC_DIR%\temp_build\cuda_11.0.2_451.48_win10.exe"
     if errorlevel 1 exit /b 1
     set "CUDA_SETUP_FILE=%SRC_DIR%\temp_build\cuda_11.0.2_451.48_win10.exe"
     set "ARGS=nvcc_11.0 cuobjdump_11.0 nvprune_11.0 cupti_11.0 cublas_11.0 cublas_dev_11.0 cudart_11.0 cufft_11.0 cufft_dev_11.0 curand_11.0 curand_dev_11.0 cusolver_11.0 cusolver_dev_11.0 cusparse_11.0 cusparse_dev_11.0 nvgraph_11.0 nvgraph_dev_11.0 npp_11.0 npp_dev_11.0 nvrtc_11.0 nvrtc_dev_11.0 nvml_dev_11.0"
 )
 
 if not exist "%SRC_DIR%\temp_build\cudnn-11.0-windows10-x64-v8.0.2.39.zip" (
-    curl -k -L https://developer.download.nvidia.com/compute/redist/cudnn/v8.0.2/cudnn-11.0-windows10-x64-v8.0.2.39.zip --output "%SRC_DIR%\temp_build\cudnn-11.0-windows10-x64-v8.0.2.39.zip"
+    curl -k -L https://ossci-windows.s3.amazonaws.com/cudnn-11.0-windows-x64-v8.0.2.39.zip --output "%SRC_DIR%\temp_build\cudnn-11.0-windows10-x64-v8.0.2.39.zip"
```
Comment (peterjc123) [inline windows/internal/cuda_install.bat:101]: typo here?

Comment (zasdfgbnm) [reply windows/internal/cuda_install.bat]: Do you mean the `windows` vs `windows10`? Looks like it is the same here: https://github.com/pytorch/pytorch/pull/42420/files#diff-3fde3b44a644efc229244fe515ea7d39R9

Comment (peterjc123) [reply windows/internal/cuda_install.bat]: Ok. But the output name is `windows10`, right? It is confusing.

**Comment by zasdfgbnm** [comment]
@peterjc123 I will reuse pytorch/pytorch#42482 by rerunning the build. This shouldn't be merged until pytorch/pytorch#42482 is green.

File: windows/build_pytorch.bat [hunk 4]
```diff
@@ -73,8 +73,7 @@ set LIB=%cd%\mkl\lib;%LIB%
 
 :: Download MAGMA Files on CUDA builds
 set MAGMA_VERSION=2.5.2
-if "%CUDA_VERSION%" == "80" set MAGMA_VERSION=2.4.0
-if "%CUDA_VERSION%" == "90" set MAGMA_VERSION=2.5.0
+if "%CUDA_VERSION%" == "110" set MAGMA_VERSION=2.5.3
```
Comment (peterjc123) [inline windows/build_pytorch.bat:76]: I think you can just update all to 2.5.3. Leaving 92 and 100 to 2.5.2.

Comment (zasdfgbnm) [reply windows/build_pytorch.bat]: fixed

File: windows/internal/cuda_install.bat (hunk 3 above)
Comment (peterjc123) [reply windows/internal/cuda_install.bat]:
```suggestion
    curl -k -L https://ossci-windows.s3.amazonaws.com/cudnn-11.0-windows-x64-v8.0.2.39.zip --output "%SRC_DIR%\temp_build\cudnn-11.0-windows-x64-v8.0.2.39.zip"
```

File: windows/internal/cuda_install.bat [hunk 5]
```diff
@@ -87,6 +88,21 @@ if not exist "%SRC_DIR%\temp_build\cudnn-10.2-windows10-x64-v7.6.5.32.zip" (
...
+
+if not exist "%SRC_DIR%\temp_build\cuda_11.0.2_451.48_win10.exe" (
+    curl -k -L https://ossci-windows.s3.amazonaws.com/cuda_11.0.2_451.48_win10.exe --output "%SRC_DIR%\temp_build\cuda_11.0.2_451.48_win10.exe"
+    if errorlevel 1 exit /b 1
+    set "CUDA_SETUP_FILE=%SRC_DIR%\temp_build\cuda_11.0.2_451.48_win10.exe"
+    set "ARGS=nvcc_11.0 cuobjdump_11.0 nvprune_11.0 cupti_11.0 cublas_11.0 cublas_dev_11.0 cudart_11.0 cufft_11.0 cufft_dev_11.0 curand_11.0 curand_dev_11.0 cusolver_11.0 cusolver_dev_11.0 cusparse_11.0 cusparse_dev_11.0 npp_11.0 npp_dev_11.0 nvrtc_11.0 nvrtc_dev_11.0 nvml_dev_11.0"
+)
+
+if not exist "%SRC_DIR%\temp_build\cudnn-11.0-windows10-x64-v8.0.2.39.zip" (
+    curl -k -L https://ossci-windows.s3.amazonaws.com/cudnn-11.0-windows10-x64-v8.0.2.39.zip --output "%SRC_DIR%\temp_build\cudnn-11.0-windows10-x64-v8.0.2.39.zip"
```
Comment (peterjc123) [inline windows/internal/cuda_install.bat:101]: So it should be `https://ossci-windows.s3.amazonaws.com/cudnn-11.0-windows-x64-v8.0.2.39.zip`.

Comment (zasdfgbnm) [reply windows/internal/cuda_install.bat:101]: Yeah, I see. I just checked our cuDNN release page, for 10.2 there is a `windows10` and `windows`, but for 11.0, there is only a `windows`. Sorry for being confusing, I will change the link and output names to `windows`.

**Comment by zasdfgbnm** [comment]
The build/install logic should be OK, but I see some compilation errors that I believe fixed in your https://github.com/pytorch/pytorch/pull/42420. I will wait for the merge of your PR and rerun the test again.

**Comment by zasdfgbnm** [comment]
@malfet This should be ready too. This is tested at https://github.com/pytorch/pytorch/pull/42482, which is failing with the same cuSparse error.

File: windows/templates/build_task.yml [hunk 6]
```diff
@@ -109,6 +109,18 @@ jobs:
         PY3.8_102:
           DESIRED_PYTHON: 3.8
           CUDA_VERSION: 102
+        PY3.5_110:
+          DESIRED_PYTHON: 3.5
+          CUDA_VERSION: 110
```
Comment (malfet) [inline windows/templates/build_task.yml:114]: We no longer support Python3.5 but I guess it can be done in a separate PR

**Pull request merged**